packages/apps/job-launcher/server/src/modules/job/job-moderation.service.ts

```diff
@@ -0,0 +1,289 @@
+import { sendSlackNotification } from '../../common/utils/slack';
+import { JobEntity } from './job.entity';
+import { JobRepository } from './job.repository';
+import { JobStatus } from '../../common/enums/job';
+import { sleep } from '../../common/utils/sleep';
+
+@Injectable()
+export class JobModerationService {
+  private visionClient: ImageAnnotatorClient;
+  public readonly logger = new Logger(JobModerationService.name);
+
+  constructor(
+    private readonly jobRepository: JobRepository,
+    private readonly visionConfigService: VisionConfigService,
+    private readonly slackConfigService: SlackConfigService,
+    private readonly storageService: StorageService,
+  ) {}
+
+  public async jobModeration(jobEntity: JobEntity): Promise<JobEntity> {
+    const manifest = (await this.storageService.downloadJsonLikeData(
+      jobEntity.manifestUrl,
+    )) as any;
+
+    const dataModerationResults: DataModerationResultDto =
+      await this.dataModeration(manifest.data.data_url);
+
+    if (dataModerationResults.containsAbuse) {
+      const abusiveImageLinks = dataModerationResults.veryLikelyOrLikelyResults
+        .map((result) => `- ${result.imageUrl}`)
+        .join('\n');
+
+      const message = `The following images contain abusive content and caused the job to fail:\n${abusiveImageLinks}`;
+
+      this.logger.error(message);
+
+      jobEntity.failedReason = message;
+      jobEntity.status = JobStatus.FAILED;
+      await this.jobRepository.updateOne(jobEntity);
+
+      return jobEntity;
+    }
+
+    // If there are possible moderation issues, save links to GCS and send a Slack notification
+    if (dataModerationResults.possibleResults.length > 0) {
+      const imageLinks = dataModerationResults.possibleResults
+        .map((result) => result.imageUrl)
+        .join('\n');
+
+      const storage = new Storage({
+        projectId: this.visionConfigService.projectId,
+        credentials: {
+          private_key: this.visionConfigService.privateKey,
+          client_email: this.visionConfigService.clientEmail,
+        },
+      });
+
+      const bucketName = this.visionConfigService.possibleResultsBucket;
+      const fileName = `moderation_results_${jobEntity.id}.txt`;
+
+      const file = storage.bucket(bucketName).file(fileName);
+      const stream = file.createWriteStream({ resumable: false });
+      stream.end(imageLinks);
+
+      const [signedUrl] = await file.getSignedUrl({
+        action: 'read',
+        expires: Date.now() + 60 * 60 * 1000, // 1 hour
+      });
+
+      const message = `The following images have possible moderation issues for job id ${jobEntity.id}. The results are saved <${signedUrl}|here>.`;
+
+      await sendSlackNotification(this.slackConfigService.webhookUrl, message);
+
+      this.logger.log(
+        'Slack notification sent with possible image links and signed URL:',
+        signedUrl,
+      );
+    }
+
+    jobEntity.status = JobStatus.MODERATION_PASSED;
```

Comment:
For possible abuse, the status should not be set to MODERATION_PASSED.
We should probably use a new status POSSIBLE_ABUSE_IN_REVIEW